fabtests/multi_recv, prov/shm: change to better test FI_MULTI_RECV and fix shm FI_MULTI_RECV completion reporting

aingerson opened the pull request:
+small change removing unused variable

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -198,29 +198,24 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
-				   struct smr_ep_entry *entry, size_t len)
+static void smr_progress_multi_recv(struct smr_ep *ep, struct smr_queue *queue,
+				    struct smr_ep_entry *entry,
+				    struct smr_ep_entry *dup_entry, size_t len)
```
The call takes a 'dup_entry' as input, but all it wants to do with it is return the flags from entry, with one of the bits cleared.  And that's only in some cases.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -381,9 +376,17 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
-			cmd->msg.hdr.op_flags | (entry->flags & ~SMR_MULTI_RECV),
-			total_len, entry->iov[0].iov_base, cmd->msg.hdr.addr,
+
+	comp_entry = *entry;
```
The ep_entry is copied here...

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -381,9 +376,17 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
+	comp_entry = *entry;
+	if (entry->flags & SMR_MULTI_RECV) {
+		smr_progress_multi_recv(ep, recv_queue, entry, &comp_entry,
+					total_len);
```
The duplicate ep_entry flags may be modifed here...

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -381,9 +376,17 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
+	if (entry->flags & SMR_MULTI_RECV) {
+		smr_progress_multi_recv(ep, recv_queue, entry, &comp_entry,
+					total_len);
+		free_entry = (comp_entry.flags == entry->flags);
```
and all we access from the duplicate are the flags here...

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -381,9 +376,17 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
+	}
+
+	ret = smr_complete_rx(ep, comp_entry.context, cmd->msg.hdr.op,
+			cmd->msg.hdr.op_flags | comp_entry.flags,
```
and here...

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -381,9 +376,17 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
+
+	ret = smr_complete_rx(ep, comp_entry.context, cmd->msg.hdr.op,
+			cmd->msg.hdr.op_flags | comp_entry.flags,
+			total_len, comp_entry.iov[0].iov_base, cmd->msg.hdr.addr,
```
The iov for the duplicate is the same as the original entry, I think.  The duplicate ep_entry doesn't seem necessary, since all this code seems to care about is whether some flag has been set or not.

The setting of free_entry = entry->flags == duplicate entry->flags is indirect.  The entry is being freed if it is not a multi-recv msg, or if it is but the space left is smaller than the minimal size.  A more explicit coding of this would be clearer, rather than hiding it behind a function and masked flags.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -651,11 +651,20 @@ int smr_progress_unexp(struct smr_ep *ep,
-			unexp_msg->cmd.msg.hdr.addr, entry->tag,
-			unexp_msg->cmd.msg.hdr.data, entry->err);
+			unexp_msg->cmd.msg.hdr.addr, comp_entry.tag,
+			unexp_msg->cmd.msg.hdr.data, comp_entry.err);
```
Similar comments as above.  comp_entry here can be replaced by entry, since the values are unchanged.  The only change is to the flags, so making a copy of entry isn't necessary.

I would change smr_progress_multi_recv() to take free_entry as input.  I would also change free_entry to a bool, since that's how it is used.

aingerson commented:
@shefty btw I've merged PR #5584 into this PR because they both have to do with MULTI_RECV. I'm going to look into the source address concerns you expressed in the other PR which I think is what's causing the issue for EFA.
psm2 is an error with number of completions. I remember chatting with you about whether we should expect one buffer to get exhausted before any others do and I think we decided it would...? does psm2 need to be fixed or should we omit that check from the test?

shefty commented:
I would expect each buffer to be consumed before moving onto the next buffer.  We may not be able to predict exactly when a buffer may be released because the provider may need to align data, but should be able to determine the max number of completions.  Is psm2 freeing the buffer 'too soon'?  That may be okay, as long as the buffer is mostly consumed.

aingerson commented:
No, the buffer is freed to late. Ie we expect to get 64 messages before the buffer is consumed but we get 127 completions and then 2 freed buffer completions.

shefty commented:
We should change psm2.  @j-xiong probably doesn't have a clue what we're talking about though.

aingerson commented:
Yeah some background would probably be helpful
@j-xiong I modified the mult_recv test to verify that the number of messages that could fit in the buffer are completed. Each message it gets, we increment the counter and then verify that we got the correct number of completions per multi recv buffer before it was freed. psm2 is reporting it got 127 messages before a buffer got freed so I suspect that the re-insertion into the recv queue is causing the two buffers to get selected back and forth instead of using the first buffer entirely before starting to consume the second. Let me know if you need more information.

j-xiong commented:
The multi-recv support in psm2 is emulated by reposting the buffer with updated address every time an message is received. If there are multiple multi-recv buffer in use, the order of reposting is not guaranteed. In this specific test case, incoming messages get into the two buffers alternatively and in the end both buffer are full and freed.

It will be very difficult (if possible at all) for the psm2 provider to ensure the first buffer be used out before data gets into the second buffer.

shefty commented:
The problem with that implementation is that the application sees no benefit of using 2 (or more) buffers.  They act as 1, with the receive queue going empty before the application can repost more.

Is this a limit of the psm2 API?

j-xiong commented:
The psm2 API doesn't have multi-recv support at all. The provider just plays tricks to provide limited support. It works well for one buffer but has issue with multiple buffers.

To fully support the multi-buffer scenario, the receive queue needs to be managed by the provider (instead of psm2 library). This would mean an extra layer of abstraction. 

Can this be an attribute that the application can query?

shefty commented:
I think we just need to document the restriction and disable the check or test for psm2.

Though I guess the provider _could_ queue multi-recv buffers internally, and not post the 2nd buffer until the 1st buffer is being released. I.e. only repost the 1st buffer until consumed.  Maybe that would work okay without much of a delay?

j-xiong commented:
That's essentially maintaining a dedicated queue for multi-recv buffers. It should work if the multi-recv buffers are not mixed with regular buffers, although normally application wouldn't mix them.

shefty commented:
Technically, you could even queue non-multi-recv buffers to the same list and hold off posting them until a multi-recv buffer at the front completed.  I'd be fine ignoring that, since mixing buffer types is unlikely to occur.

But does this sound like it would work reasonably well?  I'm assuming psm2 provides buffering in the library, which would limit data stalls. We at least wouldn't be waiting for the app to process a received message before reposting the buffer.

j-xiong commented:
Queuing all the buffers would be maintaining a full recv queue :) I would try to avoid that whenever possible because it complicates the recv path for all messages.

Maintaining a multi-recv only queue should work fine. PSM2 has internal buffering and no special action is expected from the application side.

aingerson commented:
Updated to consolidate and simplify code and reorganize patches. Also added the multi_recv test to the psm2 exclude list because of the completion checking. Let me know if you want to do something else for that.
I was also trying to understand your comment @shefty in #5584 about requiring -s and I'm not sure I follow. Even without these changes, I see tcp always failing any test if we don't have -s and with these changes I can still run this test with verbs and psm2 without -s.
AWS will still probably fail but the logs were too old for me to see the errors so I will take a look at those after this run.

shefty commented on prov/shm/src/smr_comp.c:
```diff
@@ -108,6 +108,11 @@ int smr_rx_comp(struct smr_ep *ep, void *context, uint32_t op,
+		return ofi_cq_write_overflow(ep->util_ep.rx_cq, context,
+					     smr_rx_cq_flags(op, flags),
+					     len, buf, data, tag, addr);
+
```
I understand this code change, but the commit message is confusing as heck.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -51,120 +51,92 @@ static inline uint16_t smr_convert_rx_flags(uint64_t fi_flags)
+	ret = smr_progress_unexp_queue(ep, entry, &ep->unexp_msg_queue);
+out:
+	fastlock_release(&ep->region->lock);
+	fastlock_release(&ep->util_ep.rx_cq->cq_lock);
```
Locks are being released in a different order than they were acquired.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -202,28 +202,20 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
-		return ret;
-	}
+	if (left < ep->min_multi_recv_size)
+		return 1;
```
convert from int to bool if this function is basically returning true/false

shefty replied:
Okay, you changed to bool, but still return 0 and 1, rather than true/false.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -656,29 +635,69 @@ int smr_progress_unexp(struct smr_ep *ep,
+						       &match_attr);
+		if (!dlist_entry)
+			break;
+	} while (1);
```
can write as while (dlist_entry)

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -391,23 +355,64 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
+	ret = smr_progress_msg_common(ep, cmd,
+			container_of(dlist_entry, struct smr_ep_entry, entry));
+	ofi_cirque_discard(smr_cmd_queue(ep->region));
+	return ret < 0 ? ret : 1;
```
Why 1 on success, instead of 0?

aingerson commented:
@rajachan does this test now need to be excluded for EFA since we're using the OOB_SYNC? What are your thoughts on it?

aingerson commented:
@shefty Updated with fixes and adding exclusion of the test to the EFA provider. @rajachan any thoughts on this?

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -51,120 +51,92 @@ static inline uint16_t smr_convert_rx_flags(uint64_t fi_flags)
 
 static inline struct smr_ep_entry *smr_get_recv_entry(struct smr_ep *ep,
-		fi_addr_t addr, uint64_t flags)
+		const struct iovec *iov, size_t count, fi_addr_t addr,
```
I would drop the inline from this.  The call is a little long, and the compiler may ignore it anyway.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -51,120 +51,92 @@ static inline uint16_t smr_convert_rx_flags(uint64_t fi_flags)
+		void *context, uint64_t tag, uint64_t ignore, uint64_t flags)
 {
-	struct smr_ep_entry *entry;
+	struct smr_ep_entry *entry = NULL;
```
NULL assignment is unnecessary.

shefty commented on fabtests/functional/multi_recv.c:
```diff
@@ -161,7 +164,7 @@ static int run_test(void)
 
 	if (opts.machr)
-		show_perf_mr(opts.transfer_size, opts.iterations,
+		show_perf_mr(opts.transfer_size, opts.iterations,	
```
nit: added tab at end - unneeded change

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -377,126 +349,68 @@ struct fi_ops_msg smr_msg_ops = {
 
 	dlist_insert_tail(&entry->entry, &ep->trecv_queue.list);
-	return 0;
+	ret = smr_progress_unexp_queue(ep, entry, &ep->unexp_tagged_queue);
```
The smr_generic_recv/trecv are basically the same.  You could create a single function if you pass the recv_queue and unexp_queue to it.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -333,7 +325,9 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
 	size_t total_len = 0;
-	int err, ret = 0;
+	uint16_t comp_flags;
+	int err, ret = 0, free_entry = 1;
```
I guess free_entry would be a bool as well...

rajachan commented:
Sorry, I'm behind on my Github issue/PR scan this week. We should be able to run the the multi_recv test with EFA with the OOB sync, but seemingly only when the OOB address exchange is used as well — which the current AWS-CI scripts do not. Give us a day and we'll figure out if we can fix up CI.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -656,29 +635,67 @@ int smr_progress_unexp(struct smr_ep *ep,
+	if (free_entry) {
+		dlist_remove(&entry->entry);
+		freestack_push(ep->recv_fs, entry);
+	}
```
This function is very similar to the one above it (from line 638 down, compared to line 383).  Try to merge?

rajachan commented:
Wait, OOB sync implies OOB address exchange too. Maybe I am missing why you want to exclude the test then.

aingerson commented:
> Wait, OOB sync implies OOB address exchange too. Maybe I am missing why you want to exclude the test then.

@rajachan I'm not sure what the issue with EFA is. It's failing with a bind/connect issue and the exclude file has a comment about excluding a test because it uses OOB in some sort of way but I don't understand that issue either. Just assumed they were somehow related.

rajachan commented:
Got it, so this PR had failed before you excluded it. @dancejic can you give this PR (without the test exclusion commit) a shot? I know you tested OOB sync with the multi_recv test as part of your PR https://github.com/ofiwg/libfabric/pull/5491

aingerson commented:
Updated
